### ISSUE TITLE
gp: NormalizingKalmanPrior — NKF model surface with NumPyro hyperpriors

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/pyrox#205]
## Problem / Request
 
The `gauss_flows` issues give a **density** — `Transformed(LGSSM, warp)` with a
`log_prob`. They do not give a **model**: no NumPyro sites on the state-space
parameters, no `factor` for the collapsed marginal, no path into
`pyrox.inference.EnsembleMAP` / `EnsembleVI`, no `predict`.
 
Add `NormalizingKalmanPrior` in `pyrox-gp`, mirroring the existing
`MarkovGPPrior` surface, plus a `pyrox-gp[flows]` optional extra.
 
Because `LGSSM` lives in `gaussx` (a hard `pyrox-gp` dependency), the
**unwarped** multivariate state-space model works with no flow dependencies at
all. The `flows` extra is needed only to supply a warp.
 
## User Story
 
> As a user who already fits temporal GPs with `MarkovGPPrior` and
> `markov_gp_factor`, I want the Normalizing Kalman Filter to plug into the same
> NumPyro model / guide / ensemble machinery, so that adding an observation warp
> is a one-argument change rather than a new training loop.
 
## Motivation
 
`pyrox-gp` already owns exactly this pattern for the un-warped case:
`MarkovGPPrior` (`_markov.py:146`) wraps an `SDEKernel`, exposes `log_marginal`,
and `markov_gp_factor` (`_markov.py:453`) registers it as `numpyro.factor`. That
is what lets a user put priors on lengthscale and noise, run SVI or MCMC, and use
`EnsembleMAP` over seeds. An NKF built only in `gauss_flows` gets none of it and
users would hand-roll an Optax loop.
 
**The port is cheap for a specific reason: the warp preserves conjugacy.** The
flow acts on observations, so
 
```text
log|det ∂G⁻¹/∂y_t|
```
 
does not depend on the latent state `x_t`. The Kalman recursion is therefore
still exact, the marginal likelihood is still closed-form, and the log-det is an
additive constant w.r.t. the latent path. Consequences:
 
- `numpyro.factor` works with the collapsed marginal, exactly as
  `markov_gp_factor` does.
- **None of `_inference_nongauss_markov.py` is needed** — no CVI, no posterior
  linearisation, no Gauss–Newton. A warped observation model is *not* a
  non-Gaussian likelihood in the sense that module handles.
- Ensembling over seeds is free, which matters: the SSM parameters
  `(A, H, Q, R)` are rotationally non-identifiable in the same way a latent-factor
  mixing matrix is.
Verified numerically (`nkf_verify.py` §1): `KF(G⁻¹(y)) + Σ logdet` matches the
dense `(TM, TM)` joint change-of-variables to **7.11e-15**.
 
## Proposed API
 
```python
class NormalizingKalmanPrior(eqx.Module):
    r"""Normalizing Kalman Filter prior over a multivariate time grid.
 
    Wraps a `gauss_flows.LGSSM` (or `MaskedLGSSM`) base and a
    per-timestep observation warp into the `pyrox_gp` model surface, so
    the state-space parameters and the warp can carry NumPyro priors and
    be fitted with any `pyrox.inference` driver.
 
    Because the warp acts on observations rather than on the latent state,
    the log-determinant term is independent of $x_t$ and the Kalman
    recursion stays exact -- the marginal likelihood is closed-form and no
    non-Gaussian inference strategy is required.
 
    Attributes:
        base: `gaussx.LGSSM` or `gaussx.MaskedLGSSM` with event shape
            ``(T, M)``. Available without the ``flows`` extra.
        warp: Bijection with event shape ``(M,)``, applied independently
            at each step. Defaults to identity, in which case this reduces
            exactly to a multivariate `MarkovGPPrior`.
        mean_fn: Optional callable ``times -> (T, M)``, subtracted before
            filtering and added back at predict time. Mirrors
            `MarkovGPPrior.mean_fn`.
    """
 
    base: LGSSM                      # from gaussx -- no flow deps
    warp: AbstractBijection | None = None   # from gauss_flows -- [flows] extra
    mean_fn: Callable[[Float[Array, " T"]], Float[Array, "T M"]] | None = None
 
    def log_marginal(
        self,
        y: Float[Array, "T M"],
        mask: Bool[Array, "T M"] | None = None,
    ) -> Float[Array, ""]:
        """Exact log marginal likelihood, warp included."""
 
    def predict(
        self,
        y: Float[Array, "T M"],
        mask: Bool[Array, "T M"] | None = None,
        *,
        n_ahead: int = 0,
        order: int = 32,
    ) -> tuple[Float[Array, "T M"], Float[Array, "T M"]]:
        """Predictive moments in **observation** space.
 
        Two stages: RTS smoothing gives Gaussian moments in the warped
        space, then those are pushed through $G$ by Gauss-Hermite
        quadrature. Note the mean is $E[G(z)]$, not $G(E[z])$.
        """
 
 
def normalizing_kalman_factor(
    name: str,
    prior: NormalizingKalmanPrior,
    y: Float[Array, "T M"],
    mask: Bool[Array, "T M"] | None = None,
) -> None:
    """Register the NKF marginal log-likelihood with NumPyro.
 
    Mirrors `pyrox_gp.markov_gp_factor`.
    """
    numpyro.factor(name, prior.log_marginal(y, mask))
```
 
## Design Snapshot
 
Intended usage, matching the existing `MarkovGPPrior` docstring style:
 
```python
import numpyro, numpyro.distributions as dist
from gaussx import LGSSM                      # hard dep, already present
from pyrox_gp import NormalizingKalmanPrior, normalizing_kalman_factor
 
def model(y, mask=None):
    log_q = numpyro.sample("log_q", dist.Normal(0.0, 1.0).expand([N]))
    log_r = numpyro.sample("log_r", dist.Normal(0.0, 1.0).expand([M]))
    base  = LGSSM(A, H, jnp.diag(jnp.exp(log_q)), jnp.diag(jnp.exp(log_r)),
                  m0, P0, n_steps=T)
    prior = NormalizingKalmanPrior(base, warp=warp)
    normalizing_kalman_factor("nkf", prior, y, mask)
```
 
Optional-dependency guard, following the `optax` pattern at
`_inference_nongauss.py:790` (import inside the call, not at module scope):
 
```python
def _require_flows():
    try:
        import gauss_flows
    except ImportError as exc:  # pragma: no cover
        raise ImportError(
            "NormalizingKalmanPrior requires the flows extra: "
            "pip install 'pyrox-gp[flows]'"
        ) from exc
    return gauss_flows
```
 
`pyproject.toml`:
 
```toml
[project.optional-dependencies]
# `pip install pyrox-gp[flows]` enables the Normalizing Kalman Filter and any
# other warped model surface. Pulls flowjax, paramax, scikit-learn, diffrax.
flows = [
    "gauss-flows @ git+https://github.com/jejjohnson/gauss_flows.git@<pin>",
]
```
 
Pin by commit, matching the existing `gaussx` / `geonnax` entries.
 
## Mathematical Notes
 
```text
Model:   x_t = A x_{t-1} + q_t,   y_t = G(H x_t + r_t)
 
Marginal likelihood:
 
  log p(y_1:T) = log p_LGSSM(z_1:T) + Σ_t log|det ∂G⁻¹/∂y_t|,   z_t = G⁻¹(y_t)
 
CONJUGACY. The second term depends only on y, never on x. So:
 
  ∂/∂x [ log-det term ] = 0
 
which is why the Kalman recursion is untouched and why numpyro.factor over the
COLLAPSED marginal is valid. Contrast with a warp applied to the latent state,
which would make the emission nonlinear and force EKF/UKF/CVI machinery.
 
PREDICTION is the one place quadrature enters. RTS smoothing gives
z_t ~ N(m_t, S_t) in warped space; observation-space moments are
 
  E[y_t]   = ∫ G(z) N(z; m_t, S_t) dz
  Var[y_t] = ∫ G(z)² N(z; m_t, S_t) dz − E[y_t]²
 
by Gauss-Hermite. Use gaussx's Gauss-Hermite integrator. NOTE the warp-smoothness
caveat from the TGP work: GH converges spectrally only for analytic integrands,
so a piecewise RQ-spline warp plateaus around ~3e-3 and can get WORSE with
increasing order. Order 32 is the sweet spot for splines; do not document
"raise the order until it converges" as a diagnostic.
```
 
## References & Existing Code
 
- Design doc: `kalman_survae_report.md` §1.1, §4-C/D; architecture note on layering
- Paper: de Bézenac et al., NeurIPS 2020 — [amazon.science](https://www.amazon.science/publications/normalizing-kalman-filters-for-multivariate-time-series-analysis)
- **The surface to mirror**: `packages/pyrox-gp/src/pyrox_gp/_markov.py:146` (`MarkovGPPrior`), `:376` (`ConditionedMarkovGP`), `:453` (`markov_gp_factor`), `:469` (`markov_gp_sample`)
- Optional-import precedent: `packages/pyrox-gp/src/pyrox_gp/_inference_nongauss.py:790`
- Ensemble drivers this unlocks: `packages/pyrox/src/pyrox/inference/` (`EnsembleMAP`, `EnsembleVI`, `ensemble_predict`)
- Upstream: `gauss_flows.LGSSM` / `MaskedLGSSM` / `normalizing_kalman_filter`
## Implementation Steps
 
- [ ] Add `packages/pyrox-gp/src/pyrox_gp/_markov_flow.py` with `NormalizingKalmanPrior` and `normalizing_kalman_factor`
- [ ] Add the `flows` extra to `packages/pyrox-gp/pyproject.toml`, pinned by commit
- [ ] Guard the `gauss_flows` import inside the call, with an install hint in the message
- [ ] Implement `predict` via RTS smoothing + Gauss-Hermite pushforward; default `order=32`
- [ ] Export the new names from `packages/pyrox-gp/src/pyrox_gp/__init__.py` behind the same guard
- [ ] Add both to the `members:` list in `docs/api/gp.md`
- [ ] Skip the whole test module with `pytest.importorskip("gauss_flows")` so the base install stays green
- [ ] Confirm no changes are needed in `_inference_nongauss_markov.py` — and say so in the module docstring, since "warped ⇒ non-Gaussian" is the wrong intuition here
## Definition of Done
 
- [ ] Code lands at the intended path
- [ ] Public API exported via `packages/pyrox-gp/src/pyrox_gp/__init__.py`
- [ ] Tests pass: `make test` both with and without the `flows` extra installed
- [ ] Lint + typecheck pass: `make lint && make typecheck`
- [ ] Docstrings (Google-style, `$...$` math, no Sphinx/RST) on all public symbols
## Testing
 
New file `packages/pyrox-gp/tests/gp/test_markov_flow.py`.
 
- [ ] **Reduction test:** `warp=None` makes `log_marginal` equal a direct `gaussx.kalman_filter` call on the same `(A, H, Q, R)` — exactly. This is the test that catches a mis-wired warp default
- [ ] Unit test: `log_marginal` matches `gauss_flows.normalizing_kalman_filter(...).log_prob(y)` to ≤ 1e-13
- [ ] Unit test: `log_marginal` matches the dense `(TM, TM)` joint change-of-variables to ≤ 1e-13 (measured 7.11e-15)
- [ ] Masked test: passing a `(T, M)` mask agrees with the exact marginal over observed entries to ≤ 1e-13
- [ ] NumPyro integration: `numpyro.handlers.trace` over the example model shows the expected sample sites plus one `factor` site; SVI runs 50 steps and the loss decreases
- [ ] MAP test: `EnsembleMAP` over 4 seeds runs and returns finite parameters
- [ ] Predict test: `predict` returns `E[G(z)]`, not `G(E[z])` — assert the two differ for a non-affine warp and that the quadrature mean matches a 400k-sample Monte-Carlo estimate within MC tolerance
- [ ] Predict test: with an identity warp, `predict` reduces to the RTS smoother moments exactly
- [ ] Import test: with `gauss_flows` absent, `NormalizingKalmanPrior(base, warp=None)` constructs and fits normally — only passing a `warp` raises `ImportError` with the install hint. This is the payoff of putting `LGSSM` in `gaussx`; assert it explicitly
- [ ] Docstring examples execute (the repo's doctest gate)
## Documentation
 
- [ ] `docs/api/gp.md` — add to the Markov / state-space section, stating explicitly that this needs `pyrox-gp[flows]`
- [ ] Notebook: NKF vs plain `MarkovGPPrior` on a multivariate panel with skewed marginals, showing where the warp buys calibration
- [ ] Docstrings (covered by Definition of Done)
## Relationships
 
- Parent (theme epic): #
- Blocked by: # (gaussx: LGSSM / MaskedLGSSM), # (gauss_flows: normalizing_kalman_filter + guard)
- Blocks: #
- Related: # (gaussx: per-dimension observation mask)
